packages/react/src/SelectPanel/SelectPanel.tsx

```diff
@@ -347,13 +406,32 @@ export function SelectPanel({
     return <T extends React.HTMLAttributes<HTMLElement>>(props: T) => {
       return renderAnchor({
         ...props,
-        children: selectedItems.length ? selectedItems.map(item => item.text).join(', ') : placeholder,
+        children: selectedItems.length
+          ? selectedItems
+              .sort(sortFn ?? compareByKey(sortKey, sortDirection))
+              .map(item => item.text)
+              .join(', ')
+          : placeholder,
       })
     }
-  }, [placeholder, renderAnchor, selected])
+  }, [placeholder, renderAnchor, selected, sortDirection, sortFn, sortKey])
 
-  const itemsToRender = useMemo(() => {
-    return items.map(item => {
+  useEffect(() => {
+    if (open) {
+      resetSort()
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [open])
+
+  useEffect(() => {
+    if (prevItems.length === 0 && items.length > 0) {
+      resetSort()
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [items, prevItems])
```

Comment:
Instead of synchronizing with an effect one idea is to:

```suggestion
  const [previousItems, setPreviousItems] = useState(items);
  
  if (previousItems !== items) {
    setPreviousItems(items);
    if (previousItems.length === 0 && items.length > 0) {
      resetSort()
    }
  }
```

Based on: https://react.dev/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes